ably.ts.d

```diff
@@ -0,0 +1,400 @@
+
+namespace ConnectionState {
+  export type INITIALIZED = "initialized";
+  export type CONNECTING = "connecting";
+  export type CONNECTED = "connected";
+  export type DISCONNECTED = "disconnected";
+  export type SUSPENDED = "suspended";
+  export type CLOSING = "closing";
+  export type CLOSED = "closed";
+  export type FAILED = "failed";
+}
+type ConnectionState = ConnectionState.INITIALIZED | ConnectionState.CONNECTED | ConnectionState.CONNECTING | ConnectionState.DISCONNECTED | ConnectionState.SUSPENDED | ConnectionState.CLOSED | ConnectionState.CLOSING | ConnectionState.FAILED;
+
+namespace ConnectionEvent {
+  export type INITIALIZED = "initialized";
+  export type CONNECTING = "connecting";
+  export type CONNECTED = "connected";
+  export type DISCONNECTED = "disconnected";
+  export type SUSPENDED = "suspended";
+  export type CLOSING = "closing";
+  export type CLOSED = "closed";
+  export type FAILED = "failed";
+  export type UPDATE = "update";
+}
+type ConnectionEvent = ConnectionEvent.INITIALIZED | ConnectionEvent.CONNECTED | ConnectionEvent.CONNECTING | ConnectionEvent.DISCONNECTED | ConnectionEvent.SUSPENDED | ConnectionEvent.CLOSED | ConnectionEvent.CLOSING | ConnectionEvent.FAILED | ConnectionEvent.UPDATE;
+
+namespace PresenceAction {
+  export type ABSENT = "absent";
+  export type PRESENT = "present";
+  export type ENTER = "enter";
+  export type LEAVE = "leave";
+  export type UPDATE = "update";
+}
+type PresenceAction = PresenceAction.ABSENT | PresenceAction.PRESENT | PresenceAction.ENTER | PresenceAction.LEAVE | PresenceAction.UPDATE;
+
+namespace StatsIntervalGranularity {
+  export type MINUTE = "minute";
+  export type HOUR = "hour";
+  export type DAY = "day";
+  export type MONTH = "month";
+}
+type StatsIntervalGranularity = StatsIntervalGranularity.MINUTE | StatsIntervalGranularity.HOUR | StatsIntervalGranularity.DAY | StatsIntervalGranularity.MONTH;
+
+// Interfaces
+
+interface ClientOptions extends AuthOptions {
+  autoConnect?: boolean;
+  clientId?: String;
+  defaultTokenParams?: TokenParams;
+  echoMessages?: boolean;
+  environment?: String;
+  log?: LogInfo;
+  port?: number;
+  queueMessages?: boolean;
+  restHost?: String;
+  realtimeHost?: String;
+  fallbackHosts?: Array<String>;
+  recover?: String | standardCallback;
+  tls?: boolean;
+  tlsPort?: number;
+  useBinaryProtocol?: boolean;
+}
+
+interface AuthOptions {
+  authCallback?: (data: TokenParams,callback: (Error: ErrorInfo | String, tokenRequestOrDetails: String | TokenDetails | TokenRequest) => void) => void;
+  authHeaders?: Array<any>;
+  authMethod?: String;
+  authParams?: Array<any>;
+  authUrl?: String;
+  key?: String;
+  queryTime?: boolean;
+  token?: String | TokenDetails;
+  tokenDetails?: TokenDetails;
+  useTokenAuth?: boolean;
+}
+
+interface TokenParams {
+  capability?: String;
+  clientId?: String;
+  nonce?: String;
+  timestamp?: Number;   // Time?
```

Comment:
What's the `Time?` for?  This is milliseconds since the epoch. For various reasons, we decided against a Time/Date type because of various issues across platforms and devices, whereas milliseconds since the epoch is pretty reliable.